python/lsst/cp/pipe/ptc.py

```diff
@@ -0,0 +1,535 @@
+        """
+
+        # setup necessary objects
+        detNum = dataRef.dataId[self.config.ccdKey]
+        detector = dataRef.get('camera')[dataRef.dataId[self.config.ccdKey]]
+        ampInfoCat = detector.getAmpInfoCatalog()
+        ampNames = [amp.getName() for amp in ampInfoCat]
+        dataDict = {key: {} for key in ampNames}
+        fitVectorsDict = {key: ([], [], []) for key in ampNames}
+
+        self.log.info('Measuring PTC using %s visits for detector %s' % (visitPairs, detNum))
+
+        for (v1, v2) in visitPairs:
+            # Perform ISR on each exposure
+            dataRef.dataId['visit'] = v1
+            exp1 = self.isr.runDataRef(dataRef).exposure
+            dataRef.dataId['visit'] = v2
+            exp2 = self.isr.runDataRef(dataRef).exposure
+            del dataRef.dataId['visit']
+
+            checkExpLengthEqual(exp1, exp2, v1, v2, raiseWithMessage=True)
+            expTime = exp1.getInfo().getVisitInfo().getExposureTime()
+
+            for amp in detector:
+                im1Area = exp1.maskedImage[amp.getBBox()]
+                im1Area = afwMath.binImage(im1Area, self.config.binSize)
+
+                im2Area = exp2.maskedImage[amp.getBBox()]
+                im2Area = afwMath.binImage(im2Area, self.config.binSize)
+
+                #  Clipped mean of images; then average of mean.
+                mu1 = afwMath.makeStatistics(im1Area, afwMath.MEANCLIP).getValue()
+                mu2 = afwMath.makeStatistics(im2Area, afwMath.MEANCLIP).getValue()
+                mu = 0.5*(mu1 + mu2)
+
+                # Take difference of pairs
+                # symmetric formula: diff = (mu2*im1-mu1*im2)/(0.5*(mu1+mu2))
+                temp = im2Area.clone()
+                temp *= mu1
+                diffIm = im1Area.clone()
+                diffIm *= mu2
+                diffIm -= temp
+                diffIm /= mu
+
+                varDiff = 0.5*(afwMath.makeStatistics(diffIm, afwMath.VARIANCECLIP).getValue())
+
+                data = dict(expTime=expTime, meanClip=mu, varClip=varDiff)
+
+                ampName = amp.getName()
+                dataDict[ampName][(v1, v2)] = data
+
+                fitVectorsDict[ampName][0].append(expTime)
+                fitVectorsDict[ampName][1].append(mu)
+                fitVectorsDict[ampName][2].append(varDiff)
+
+        # Fit PTC and (non)linearity of signal vs time curve
+        fitPtcDict, nlDict = self.fitPtcAndNl(fitVectorsDict, ptcFitType=self.config.ptcFitType)
+        finalDict = {"data": dataDict, "ptc": fitPtcDict, "nl": nlDict}
+
+        # Save data, PTC fit, and NL fit dictionaries
+        self.log.info(f"Writing PTC and NL data to {dataRef.getUri(write=True)}")
+        dataRef.put(finalDict)
+
+        if self.config.makePlots:
+            self.plot(dataRef, fitPtcDict, nlDict, ptcFitType=self.config.ptcFitType)
+
+        self.log.info('Finished measuring PTC for in detector %s' % detNum)
+
```

Comment:
You are probably right and we most likely don't need all of that. I can change the output to be just the noise, the gain, and the NL threshold (what do you have in mind for this? The signal value for which the nonlinearity residual is above a certain threshold (e.g., 1%))?  @mfisherlevine : any other information that needs to be output by this task in this ticket?

Reply:
Could you summarise what the current data structure actually contains right now, and what is being suggested to save me reading the code and working it all out? Then I can easily help decide what we do/don't want.

Reply:
I changed the code to save a gain dictionary (gain, gain error per amp) and a noise dictionary (noise, noise error per amp). I also plot mean vs time and calculate a non-linearity residual that gets saved in a dictionary of the form:
nlDict[amp] = (timeVecFinal, meanVecFinal, linResidual, parsFit, parsFitErr)

Reply:
I think that all sounds worth saving, but that's how it is at the moment, right? What is being proposed to be removed or added here?

Reply:
Originally, the code was saving all the data necessary for the PTC fit (i.e., the mean vector and the variance vector per amp) in addition to all the parameters from the fit (i.e., not only gain and error, but other parameters such as a00 in the case of the Astier approximation or higher order terms, for a polynomial).

Reply:
Well, that doesn't sound like too much really. On the other hand, if it's all in one giant ugly data structure then that doesn't sound very good either, so I think there are two decent options here.

Option 1 (to be done if running the task is slow) - create a new data product that gets persisted with all that other stuff in it, so the stuff we really want is separate, but we don't lose the other valuable stuff that gets calculated

Option 2 (if the task is fast) - Just throw out all the other stuff, because it's not too slow to just rerun.

Reply:
For option 1, what other data structure? I was using a dictionary of dictionaries, which sounds good to me.

Reply:
No no, a new data _product_, _i.e._ a new one, separate from the current {{measurePhotonTransferCurveDataset}}, so that it can be saved to a different file.

Reply:
I have added a new data product as suggested by Merlin